atlp-rwanda/stackup2-barefoot-backend
# #171380759 Avail requests for approval

#### What does this PR do?
* Add `Approve trip request` functionality
* Add `Reject trip request` functionality
* Add `Reassign trip request` functionality
* Update user profile by setting line manager
#### Description of Task to be completed?
* Write endpoints unit tests
* Create controller functions
* Create helper functions, validations, middlewares, and custom messages.
* Create `approve`, `reject`, and `reassign` routes
#### How should this be manually tested?
1. Open your terminal and run `git clone https://github.com/Stackup-Rwanda/stackup2-barefoot-backend/`
2. run `cd stackup2-barefoot-backend` to go into the root directory of the project
3. run `git checkout -b ft-avail-requests-for-approval-171380759 origin/ft-avail-requests-for-approval-171380759` to pull this branch and switch to it
4. run `npm install` to install the dependencies
5. In .env file, add these variables:
SUPER_ADMIN_EMAIL = #superadmin@admin.com
SUPER_ADMIN_PASSWORD = #SuperUserPassword
* run `npm run migrations`
* run `npm run admin` to create a super admin account
* run `npm run dev`
6. Using POSTMAN, create 3 different accounts and verify them, then using super admin token, set one account 1 and 2 have a role of manager
7. Use the remaining account to create a trip request

**Approve a trip request**
1. As a requester, update your profile and set a line manager:
PATCH `http://localhost:<port>/api/profile`
Body request:
```
{
 lineManager: id
}
```
2. Login as the manager who just got a direct report above
3. Approve/reject your direct report's trip request: 
  PATCH `http://localhost:<port>/api/trips/<tripRequestId>/approve`

**Reject a trip request**
  PATCH `http://localhost:<port>/api/trips/<tripRequestId>/reject`

**Reassign a trip request to another manager**
  PATCH `http://localhost:<port>/api/trips/<tripRequestId>/reassign`
Body request:
```
{
  "userId": "id"
}
```
#### Any background context you want to provide?
* A manager can only approve or reject trip requests from their direct report and the ones assigned to him/her
* Only managers can reassign requests to other managers
* An already approved or rejected request cannot be reassigned
#### What are the relevant pivotal tracker stories?
[#171380759](https://www.pivotaltracker.com/story/show/171380759)
[#171380758](https://www.pivotaltracker.com/story/show/171380758)
#### Screenshots (if appropriate)
![Screenshot from 2020-04-02 07-43-05](https://user-images.githubusercontent.com/45935765/78215241-5c06ad00-74b7-11ea-8b68-3850df231e75.png)
![Screenshot from 2020-04-02 07-43-49](https://user-images.githubusercontent.com/45935765/78215254-6032ca80-74b7-11ea-883a-22c7a6da964c.png)
![Screenshot from 2020-04-02 07-45-43](https://user-images.githubusercontent.com/45935765/78215260-6163f780-74b7-11ea-8c49-bac60f98b875.png)

#### Questions:
N/A
